src/ovs/aca_vlan_manager.cpp

```diff
@@ -0,0 +1,205 @@
+// Copyright 2019 The Alcor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "aca_log.h"
+#include "aca_vlan_manager.h"
+#include <errno.h>
+#include <algorithm>
+
+namespace aca_vlan_manager
+{
+ACA_Vlan_Manager &ACA_Vlan_Manager::get_instance()
+{
+  // Instance is destroyed when program exits.
+  // It is instantiated on first use.
+  static ACA_Vlan_Manager instance;
+  return instance;
+}
+
+// unsafe function, needs to be called inside vpc_table_mutex lock
+bool ACA_Vlan_Manager::check_entry_existed_unsafe(string vpc_id)
+{
+  ACA_LOG_DEBUG("ACA_Vlan_Manager::check_entry_existed_unsafe ---> Entering\n");
+
+  std::unordered_map<std::string, values>::const_iterator found = vpc_table.find(vpc_id);
+
+  // found would be equal to vpc_table.end() if entry is not there
+  bool entry_existed = (found != vpc_table.end());
+
+  ACA_LOG_DEBUG("ACA_OVS_Programmer::check_entry_existed_unsafe <--- Exiting\n");
+
+  return entry_existed;
+}
+
+// unsafe function, needs to be called inside vpc_table_mutex lock
+// this function assumes there is no existing entry for vpc_id
+void ACA_Vlan_Manager::create_entry_unsafe(string vpc_id)
+{
+  ACA_LOG_DEBUG("ACA_Vlan_Manager::create_entry_unsafe ---> Entering\n");
+
+  values new_values;
+  new_values.vlan_id = current_available_vlan_id.load();
+  current_available_vlan_id++;
+
+  vpc_table.emplace(vpc_id, new_values);
+
+  ACA_LOG_DEBUG("ACA_OVS_Programmer::create_entry_unsafe <--- Exiting\n");
+}
+
+uint ACA_Vlan_Manager::get_or_create_vlan_id(string vpc_id)
+{
+  ACA_LOG_DEBUG("ACA_Vlan_Manager::acquire_vlan_id ---> Entering\n");
+
+  // -----critical section starts-----
+  vpc_table_mutex.lock();
+  if (!check_entry_existed_unsafe(vpc_id)) {
+    create_entry_unsafe(vpc_id);
+  }
+  uint acquired_vlan_id = vpc_table[vpc_id].vlan_id;
+  vpc_table_mutex.unlock();
+  // -----critical section ends-----
+
+  ACA_LOG_DEBUG("ACA_OVS_Programmer::acquire_vlan_id <--- Exiting\n");
+
+  return acquired_vlan_id;
+}
+
+void ACA_Vlan_Manager::add_ovs_port(string vpc_id, string ovs_port)
+{
+  ACA_LOG_DEBUG("ACA_Vlan_Manager::add_ovs_port ---> Entering\n");
+
+  // -----critical section starts-----
+  vpc_table_mutex.lock();
+  if (!check_entry_existed_unsafe(vpc_id)) {
+    create_entry_unsafe(vpc_id);
+  }
+  vpc_table[vpc_id].ovs_ports.push_back(ovs_port);
```

Comment:
Is it possible that push_back could bring duplicated ovs_ports, if multiple duplicated calls from upstream controller?

Reply:
Note: ovs_port name has to be unique on the host.

if ACA receive duplicate port create operation with the same port name, my thinking is ACA would return error for that operation. I have put "todo" in the code the handle that. Let me know if we want ACA to try to limp along to handle it.

Reply:
Ideally controller should block the duplicated operations for the same port. However there might be some corner cases where one port name is programmed twice to ACA so let us plan to handle it gracefully. 

Options include
- Reject call with error message
- No ops with warning message
- Do it regardless => if there is anything wrong, controller's fault :-)

Reply:
Agreed. There are pros and cons for the options. I have added them into the "todo" in the code. We can close on it later as it is not a priority for the current release.